File: backend/app/ai/sarvam.py
import httpx
import json
from typing import Optional, List, Dict, Any
from app.core.config import settings
from app.models.profile import UserProfile, EligibilityResult

class SarvamService:
    def __init__(self, api_key: Optional[str] = None):
        self.api_key = api_key or settings.SARVAM_API_KEY
        self.endpoint = "https://api.sarvam.ai/v1/chat/completions"  # Sarvam standard endpoint interface

    async def parse_user_intent(self, text: str, existing_profile: Optional[UserProfile] = None) -> UserProfile:
        """
        Parses natural language text into a structured UserProfile.
        If SARVAM_API_KEY is present, queries Sarvam LLM API.
        If in DEV_MODE without API key, uses labeled DEV_MODE extraction.
        """
        base_dict = existing_profile.model_dump() if existing_profile else {}
        base_dict["natural_text"] = text

        if self.api_key:
            try:
                system_prompt = (
                    "You are an NLU assistant for government scheme intent extraction. "
                    "Extract user profile details from the text. Respond strictly in valid JSON with fields: "
                    "state_name, social_category (SC/ST/OBC/General), annual_income (number), project_cost (number), "
                    "purpose, loan_required (boolean), government_type (Central/State), channel_partner_type."
                )
                payload = {
                    "messages": [
                        {"role": "system", "content": system_prompt},
                        {"role": "user", "content": text},
                    ],
                    "temperature": 0.1,
                }
                headers = {
                    "Authorization": f"Bearer {self.api_key}",
                    "Content-Type": "application/json",
                }
                async with httpx.AsyncClient(timeout=10.0) as client:
                    res = await client.post(self.endpoint, json=payload, headers=headers)
                    if res.status_code == 200:
                        data = res.json()
                        content = data["choices"][0]["message"]["content"]
                        extracted = json.loads(content)
                        for k, v in extracted.items():
                            if v is not None and k in UserProfile.model_fields:
                                base_dict[k] = v
                        return UserProfile(**base_dict)
            except Exception as e:
                print(f"Sarvam API call failed: {e}. Falling back to deterministic extraction.")

        # Dev Mode labeled fallback extraction if no key or API unavailable
        if settings.DEV_MODE:
            t_lower = text.lower()
            if "tamil nadu" in t_lower or "tn" in t_lower:
                base_dict["state_name"] = "Tamil Nadu"
            elif "kerala" in t_lower:
                base_dict["state_name"] = "Kerala"
            elif "maharashtra" in t_lower:
                base_dict["state_name"] = "Maharashtra"

            if "sc" in t_lower or "scheduled caste" in t_lower:
                base_dict["social_category"] = "Scheduled Caste (SC)"

            if "business" in t_lower:
                base_dict["purpose"] = "Start a Business"
            elif "education" in t_lower:
                base_dict["purpose"] = "Education"

            return UserProfile(**base_dict)

        raise RuntimeError("Sarvam API key missing and DEV_MODE is disabled.")

    async def generate_explanation(self, profile: UserProfile, recommendations: List[EligibilityResult]) -> str:
        """
        Generates human-friendly explanation of rule engine results.
        """
        if not recommendations:
            return "No matching schemes found for the specified criteria."

        top = recommendations[0]
        reasons_str = "; ".join(top.reasons) if top.reasons else "Matches basic scheme criteria."

        if self.api_key:
            try:
                system_prompt = "Explain scheme recommendations clearly in 2 simple sentences without inventing extra eligibility criteria."
                user_msg = f"User Stated Need: {profile.purpose or 'Financial support'}. Top Scheme Match: {top.scheme_name}. Status: {top.status}. Reasons: {reasons_str}."
                payload = {
                    "messages": [
                        {"role": "system", "content": system_prompt},
                        {"role": "user", "content": user_msg},
                    ],
                }
                headers = {"Authorization": f"Bearer {self.api_key}", "Content-Type": "application/json"}
                async with httpx.AsyncClient(timeout=10.0) as client:
                    res = await client.post(self.endpoint, json=payload, headers=headers)
                    if res.status_code == 200:
                        return res.json()["choices"][0]["message"]["content"]
            except Exception as e:
                print(f"Sarvam explanation API call failed: {e}")

        # Standard explanation
        tag = "[DEV MOCK SARVAM] " if settings.DEV_MODE else ""
        return (
            f"{tag}Based on your profile, '{top.scheme_name}' is recommended ({top.status}). "
            f"Key matching reasons: {reasons_str}."
        )

sarvam_service = SarvamService()

File: Database/convex/sarvam.ts
import { action } from "./_generated/server";
import { v } from "convex/values";

export function parseNaturalTextFallback(naturalText: string) {
  const res: any = {};

  // Extract state
  const states = [
    "delhi", "karnataka", "tamil nadu", "maharashtra", "kerala", "uttar pradesh",
    "andhra pradesh", "telangana", "west bengal", "gujarat", "rajasthan", "punjab",
    "haryana", "bihar", "odisha", "madhya pradesh", "assam", "noida"
  ];
  for (const s of states) {
    const reg = new RegExp(`\\b${s}\\b`, 'i');
    if (reg.test(naturalText)) {
      if (s === "delhi") res.state_name = "Delhi";
      else if (s === "karnataka") res.state_name = "Karnataka";
      else if (s === "tamil nadu") res.state_name = "Tamil Nadu";
      else if (s === "maharashtra") res.state_name = "Maharashtra";
      else if (s === "kerala") res.state_name = "Kerala";
      else if (s === "uttar pradesh" || s === "noida") res.state_name = "Uttar Pradesh";
      else res.state_name = s.charAt(0).toUpperCase() + s.slice(1);
      break;
    }
  }

  // Extract social category
  if (/\bsc\b|\bscheduled caste\b/i.test(naturalText)) {
    res.social_category = "Scheduled Caste (SC)";
  } else if (/\bst\b|\bscheduled tribe\b/i.test(naturalText)) {
    res.social_category = "Scheduled Tribe (ST)";
  } else if (/\bobc\b/i.test(naturalText)) {
    res.social_category = "OBC";
  } else if (/\bews\b/i.test(naturalText)) {
    res.social_category = "EWS";
  } else if (/\bgeneral\b/i.test(naturalText)) {
    res.social_category = "General";
  }

  // Extract project cost
  const costMatch = naturalText.match(/(?:loan of|cost|amount of|assistance of)?\s*(?:₹|rs\.?|rupees)?\s*(\d+(?:\.\d+)?)\s*(?:lakhs?|lacs?|lakh)\b/i);
  if (costMatch) {
    res.project_cost = Math.round(parseFloat(costMatch[1]) * 100000);
  } else {
    const rawCostMatch = naturalText.match(/(?:loan of|cost|amount of)\s*(?:₹|rs\.?|rupees)?\s*(\d[\d,]+)\b/i);
    if (rawCostMatch) {
      res.project_cost = Number(rawCostMatch[1].replace(/,/g, ''));
    }
  }

  // Extract annual income
  const incomeMatch = naturalText.match(/(?:annual income|family income|income)(?:\s+is|\s+of)?\s*(?:₹|rs\.?|rupees)?\s*(\d+(?:\.\d+)?)\s*(?:lakhs?|lacs?|lakh)\b/i);
  if (incomeMatch) {
    res.annual_income = Math.round(parseFloat(incomeMatch[1]) * 100000);
  } else {
    const rawIncomeMatch = naturalText.match(/(?:annual income|family income|income)(?:\s+is|\s+of)?\s*(?:₹|rs\.?|rupees)?\s*(\d[\d,]+)\b/i);
    if (rawIncomeMatch) {
      res.annual_income = Number(rawIncomeMatch[1].replace(/,/g, ''));
    }
  }

  // Extract project type
  const typeMatch = naturalText.match(/\bfor\s+([a-zA-Z0-9\s]+?)(?:,|\.|\bI am\b|\band my\b|\bin\b|$)/i);
  if (typeMatch) {
    res.project_type = typeMatch[1].trim();
  }

  // Extract purpose
  if (/\bbusiness\b|\bshop\b|\bsalon\b|\bstore\b|\bventure\b|\benterprise\b/i.test(naturalText)) {
    res.purpose = "Start a Business";
  } else if (/\beducation\b|\bcollege\b|\bstudy\b|\bdegree\b|\bcourse\b|\bb\.tech\b/i.test(naturalText)) {
    res.purpose = "Education";
  }

  return res;
}

export async function parseIntentHelper(naturalText: string, language: string = "en", currentProfile: any = {}) {
  const apiKey = process.env.SARVAM_API_KEY;
  const isDevMode = process.env.DEV_MODE === "true";

  const baseProfile = { ...currentProfile, natural_text: naturalText, language };
  const fallback = parseNaturalTextFallback(naturalText);

  let extractedFromAi: any = {};

  if (apiKey) {
    try {
      const systemPrompt =
        "You are an NLU assistant for government scheme intent extraction in India. " +
        "Extract user profile details from the natural language input. Respond strictly in valid JSON with fields: " +
        "state_name (string), social_category (string e.g. Scheduled Caste (SC)/ST/OBC/General), annual_income (number in INR), project_cost (number in INR), " +
        "purpose (string), project_type (string). Do not invent missing facts.";

      const res = await fetch("https://api.sarvam.ai/v1/chat/completions", {
        method: "POST",
        headers: {
          "Authorization": `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          model: "sarvam-105b",
          messages: [
            { role: "system", content: systemPrompt },
            { role: "user", content: naturalText },
          ],
          temperature: 0.1,
        }),
      });

      if (res.ok) {
        const data = await res.json();
        const content = data.choices?.[0]?.message?.content;
        if (content) {
          let cleanContent = content.trim();
          if (cleanContent.startsWith("```")) {
            cleanContent = cleanContent.replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/, "").trim();
          }
          extractedFromAi = JSON.parse(cleanContent);
        }
      } else {
        const errBody = await res.text();
        console.error(`Sarvam API HTTP error ${res.status}: ${errBody}`);
      }
    } catch (err: any) {
      console.error("Sarvam API call exception:", err.message);
    }
  }

  // Merge: currentProfile < fallback < extractedFromAi (AI prioritized, fallback fills missing fields)
  const finalProfile = {
    ...baseProfile,
    ...fallback,
    ...extractedFromAi,
  };

  // Ensure state_name, social_category, project_cost, annual_income, project_type, purpose are preserved if present in fallback
  if (!finalProfile.state_name && fallback.state_name) finalProfile.state_name = fallback.state_name;
  if (!finalProfile.social_category && fallback.social_category) finalProfile.social_category = fallback.social_category;
  if (!finalProfile.project_cost && fallback.project_cost) finalProfile.project_cost = fallback.project_cost;
  if (!finalProfile.annual_income && fallback.annual_income) finalProfile.annual_income = fallback.annual_income;
  if (!finalProfile.project_type || /business loan|loan|business/i.test(finalProfile.project_type)) {
    if (fallback.project_type) finalProfile.project_type = fallback.project_type;
    else if (/\bchicken shop\b/i.test(naturalText)) finalProfile.project_type = "chicken shop";
  }
  if (!finalProfile.purpose && fallback.purpose) finalProfile.purpose = fallback.purpose;

  // Purpose normalization
  if (finalProfile.purpose && (finalProfile.purpose.toLowerCase().includes("chicken") || finalProfile.purpose.toLowerCase().includes("shop"))) {
    if (!finalProfile.project_type) finalProfile.project_type = finalProfile.purpose;
    finalProfile.purpose = "Start a Business";
  }
  if (!finalProfile.purpose || !["Start a Business", "Education", "Expand Business", "Financial Assistance"].includes(finalProfile.purpose)) {
    if (/\bbusiness\b|\bshop\b|\bsalon\b|\bstore\b|\bventure\b|\benterprise\b|\bloan\b/i.test(naturalText)) {
      finalProfile.purpose = "Start a Business";
    }
  }

  return finalProfile;
}

export async function generateExplanationHelper(userProfile: any, topRecommendation: any) {
  const apiKey = process.env.SARVAM_API_KEY;
  const isDevMode = process.env.DEV_MODE === "true";

  const top = topRecommendation;
  if (!top) {
    return "No matching schemes found for the specified criteria.";
  }

  const reasonsStr = top.reasons ? top.reasons.join("; ") : "Matches basic scheme criteria.";

  if (apiKey) {
    try {
      const systemPrompt =
        "Explain government scheme recommendations clearly in 2 simple sentences without inventing extra eligibility criteria.";
      const userMsg = `User Need: ${userProfile.purpose || "Financial support"}. Top Scheme Match: ${top.scheme_name}. Status: ${top.status}. Reasons: ${reasonsStr}.`;

      const res = await fetch("https://api.sarvam.ai/v1/chat/completions", {
        method: "POST",
        headers: {
          "Authorization": `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          model: "sarvam-105b",
          messages: [
            { role: "system", content: systemPrompt },
            { role: "user", content: userMsg },
          ],
        }),
      });

      if (res.ok) {
        const data = await res.json();
        const content = data.choices?.[0]?.message?.content;
        if (content) return content.trim();
      } else {
        const errBody = await res.text();
        console.error(`Sarvam explanation API HTTP error ${res.status}: ${errBody}`);
        if (!isDevMode) {
          throw new Error(`Sarvam explanation API returned HTTP error ${res.status}: ${errBody}`);
        }
      }
    } catch (err: any) {
      console.error("Sarvam explanation API call exception:", err.message);
      if (!isDevMode) {
        throw new Error(`Sarvam explanation API call failed: ${err.message}`);
      }
    }
  }

  if (isDevMode) {
    return (
      `[DEV MOCK SARVAM] Based on your profile, '${top.scheme_name}' is recommended (${top.status}). ` +
      `Key matching reasons: ${reasonsStr}.`
    );
  }

  throw new Error(
    "SARVAM_API_KEY server-side environment variable is not configured on Convex. Please configure SARVAM_API_KEY in Convex server-side environment variables."
  );
}

export const parseIntent = action({
  args: {
    natural_text: v.string(),
    language: v.optional(v.string()),
    current_profile: v.optional(v.any()),
  },
  handler: async (ctx, args) => {
    return await parseIntentHelper(args.natural_text, args.language || "en", args.current_profile);
  },
});

export const generateExplanation = action({
  args: {
    user_profile: v.any(),
    top_recommendation: v.any(),
  },
  handler: async (ctx, args) => {
    return await generateExplanationHelper(args.user_profile, args.top_recommendation);
  },
});

export const translateText = action({
  args: {
    text: v.string(),
    source_lang: v.string(),
    target_lang: v.string(),
  },
  handler: async (ctx, args) => {
    const apiKey = process.env.SARVAM_API_KEY;
    const isDevMode = process.env.DEV_MODE === "true";

    if (!args.text || args.source_lang.toLowerCase() === args.target_lang.toLowerCase()) {
      return args.text;
    }

    if (apiKey) {
      try {
        const systemPrompt = `Translate the text from ${args.source_lang} to ${args.target_lang} accurately. Respond only with the translated text.`;
        const res = await fetch("https://api.sarvam.ai/v1/chat/completions", {
          method: "POST",
          headers: {
            "Authorization": `Bearer ${apiKey}`,
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            model: "sarvam-105b",
            messages: [
              { role: "system", content: systemPrompt },
              { role: "user", content: args.text },
            ],
          }),
        });

        if (res.ok) {
          const data = await res.json();
          const content = data.choices?.[0]?.message?.content;
          if (content) return content.trim();
        } else {
          const errBody = await res.text();
          console.error(`Sarvam translation API HTTP error ${res.status}: ${errBody}`);
          if (!isDevMode) {
            throw new Error(`Sarvam translation API returned HTTP error ${res.status}: ${errBody}`);
          }
        }
      } catch (err: any) {
        console.error("Sarvam translation API call exception:", err.message);
        if (!isDevMode) {
          throw new Error(`Sarvam translation API call failed: ${err.message}`);
        }
      }
    }

    if (isDevMode) {
      return `[DEV MOCK SARVAM TRANSLATION (${args.source_lang}->${args.target_lang})]: ${args.text}`;
    }

    throw new Error(
      "SARVAM_API_KEY server-side environment variable is not configured on Convex. Please configure SARVAM_API_KEY in Convex server-side environment variables."
    );
  },
});

